tutorials/data-science-preprocessing/index.md

```diff
@@ -0,0 +1,437 @@
+[filebasedsource]: /dataflow/model/custom-io-python#convenience-source-base-classes
+[rangetracker]: /dataflow/model/custom-io-python#implementing-the-rangetracker-subclass
+
+### Attach the functions in series
+
+Now that you've defined the source of data, you can now pipe it through to the
+filtering functions, in series. Apache Beam uses the pipe operator (`|`) to do
+this (with the double angle bracket operator (`>>`) to add a description):
+
+[embedmd]:# (src/preprocessing/clean.py /def main/ / p\.run.*/)
+```py
+def main(src_path, dest_table, pipeline_args):
+    p = apache_beam.Pipeline(argv=pipeline_args)
+
+    value = p | 'Read JSON' >> apache_beam.Read(JsonFileSource(src_path))
+
+    value |= (
+        'Remove records that lack location or year data' >>
+        apache_beam.FlatMap(discard_incomplete))
+
+    value |= (
+        'Convert string values to their types' >>
+        apache_beam.Map(convert_types))
+
+    value |= (
+        'Filter bad data' >>
+        apache_beam.FlatMap(filter_suspicious))
+
+    value |= (
+        'Massage fields with "rec" prefix' >>
+        apache_beam.Map(massage_rec))
+
+    value |= (
+        'Dump data to BigQuery' >>
+        apache_beam.Write(apache_beam.io.BigQuerySink(
+            dest_table,
+            schema=', '.join([
+                'fall:STRING',
+                'year:INTEGER',
+                'nametype:STRING',
+                'mass:FLOAT',
+                'name:STRING',
+                'class:STRING',
+                'latitude:FLOAT',
+                'longitude:FLOAT',
+                'id:STRING',
+            ]),
+            create_disposition=(
+                apache_beam.io.BigQueryDisposition.CREATE_IF_NEEDED),
+            write_disposition=(
+                apache_beam.io.BigQueryDisposition.WRITE_TRUNCATE))))
+
+    p.run()
+```
+
+Note that this example takes advantage of the built-in [`BigQuerySink`][bq-sink]
+to output the result into BigQuery, to analyze later.
+
+[bq-sink]: https://github.com/apache/incubator-beam/tree/python-sdk/sdks/python#bigquery
+
+## Run the data cleansing pipeline
+
+A great thing about Apache Beam and Cloud Dataflow is that you can run it
+locally for sample datasets, or datasets that are otherwise relatively small.
+But then you can take the same code, add a couple extra flags, and run it using
+Cloud Dataflow which can handle orders of magnitude more data at a time.
+
+Because this sample uses a `BigQuerySink` for its output, you must first perform
+some setup:
+
+* First, make sure you're using the correct project:
+
+      $ gcloud config set project your-project-id
+      Updated property [core/project]
+      $
+
+* Now, create a BigQuery dataset using the `bq` command provided by the Google
+  Cloud SDK:
+
+      $ bq mk meteor_dataset
+      Dataset 'your-project-id:meteor_dataset' successfully created
+      $
+
+* To ensure that this example's dependencies don't conflict with any existing
```

Comment:
I know you put having virtualenv installed in the prerequisites, but I wonder whether you think your audience is likely to already know about it and what it does. It seems like a good bet that an experienced Python dev would already be familiar, but what about the data scientist who's somewhat familiar with Python, but not yet adept?